### ISSUE TITLE
How to handle super-large queries

### ISSUE THREAD

**Issue by stites** [sentenai/py-sentenai#69]
A user might still, after we implement all of the responsive issues / discussions I've been making (#66, #67, #68, #65) request a super-large query. In this case, we should still try to make the experience as painless as possible. Here are some things I can think of which might help:
- Return Dask arrays #66 by default instead of numpy arrays. This is relevant because super-large queries will return in CPU-bound, single-threaded process (not ideal) and these kinds of dataframes are not very responsive. In contrast, a single-node dask array is concurrent by default.
- return an initial `head(10)` result with every query -- this might not be the best since we would send the first ten values twice.
- somehow show a lazy view of the first 10/100/1000 rows in a dataframe and continue downloading in a background thread. I'm not even sure this is possible, but allowing a user to work against a sample of the data before crunching against the entire dataset does sound nice.
- Offer downsampling techniques. This would require a bit of research on our side as to what is sane / reasonable. Offering something like "return every other event" and letting users interpolate the results, client-side, would cut download times down by half.

**Comment by mitchellwrosen** [comment]
> return an initial head(10) result with every query -- this might not be the best since we would send the first ten values twice.

This unfortunately isn't as easy as I'd like it to be on the back end. A query can involve multiple different streams, each of which are fetched in parallel, and streamed directly to the client (i.e. out of order). To  global ordering on the server would be a lot more bookkeeping, and was punted on for now (perhaps indefinitely).

How valuable is actually seeing the first few events? If "very" - we should revisit the event fetching interface.

**Comment by stites** [comment]
> How valuable is actually seeing the first few events? If "very" - we should revisit the event fetching interface.

I think these are all "nice-to-haves" for the time being. The minimal requirement here exists somewhere in #67 (progress bars) and the upgrade of infrastructure on the devops side (as brendan mentioned).

RE: the streaming interface -- yeah, the tricky part here is designing something on the python side.
